File: csharp/11_Pull/ValuesOfMonth.cs
using System;
using System.Collections.Generic;
using Common;

namespace Pull.Months
{
    public class ValuesOfMonth : IValuesOfMonth
    {
        private readonly Month month;
        private readonly IValuesOfMonth precedingMonth;
        private readonly IList<Transaction> transactions = new List<Transaction>();

        public virtual int Balance
        {
            get
            {
                int balance = precedingMonth.Balance;
                foreach (Transaction transaction in transactions)
                {
                    balance += transaction.Amount;
                }
                return balance;
            }
        }

        public virtual int AverageBalance
        {
            get
            {
                double averageBalance = precedingMonth.Balance;
                foreach (Transaction transaction in transactions)
                {
                    averageBalance += RateOf(transaction);
                }
                return (int)averageBalance;
            }
        }

        public Month Month
        {
            get { return month; }
        }

        public ValuesOfMonth(DateTime dateOfMonth, IValuesOfMonth precedingMonth)
        {
            this.month = new Month(dateOfMonth);
            this.precedingMonth = precedingMonth;
        }

        public void AddTransaction(Transaction transaction)
        {
            this.transactions.Add(transaction);
        }

        private double RateOf(Transaction transaction)
        {
            int countingDays = month.DaysInMonth - transaction.Date.Day + 1;
            double rate = (double)countingDays / month.DaysInMonth;
            return transaction.Amount * rate;
        }
    }
}

File: java/11Pull/src/pull/months/ValuesOfMonth.java
package pull.months;

import java.util.ArrayList;
import java.util.List;

import org.joda.time.LocalDate;
import org.joda.time.YearMonth;
import org.joda.time.chrono.ISOChronology;

import common.Transaction_API;

public class ValuesOfMonth implements IValuesOfMonth
{

	protected final YearMonth yearMonth;
	protected final IValuesOfMonth precedingMonth;
	protected final List<Transaction_API> transactions = new ArrayList<Transaction_API>();

	public ValuesOfMonth(LocalDate date, IValuesOfMonth month)
	{
		super();
		this.precedingMonth = month;
		this.yearMonth = new YearMonth(date);
	}

	public YearMonth getYearMonth()
	{
		return yearMonth;
	}

	public void addTransaction(Transaction_API transaction)
	{
		transactions.add(transaction);
	}

	@Override
	public int getBalance()
	{
		int result = precedingMonth.getBalance();
		for (Transaction_API transaction : transactions)
		{
			result += transaction.getAmount();
		}
		return result;
	}

	public int getAverageBalance()
	{
		int result = precedingMonth.getBalance();
		for (Transaction_API transaction : transactions)
		{
			result += calculateProportion(transaction);
		}

		return result;
	}

	private double calculateProportion(Transaction_API transaction)
	{
		int countingDays = daysInMonth() - transaction.getDate().getDayOfMonth() + 1;
		double rate = (double) countingDays / daysInMonth();
		return (transaction.getAmount() * rate);
	}

	private int daysInMonth()
	{
		return ISOChronology.getInstance().dayOfMonth().getMaximumValue(yearMonth);
	}

}
